v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/transform/AccumulatingTableReader.java

```diff
@@ -0,0 +1,125 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.Map.Entry;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+
+/**
+ * Helper to {@link ReaderTransform} that expands the Pipeline Graph to flatten all the tableReaders
+ * and generate table completions. Please refer to {@link ReaderTransform} for greater details.
+ */
+@AutoValue
+abstract class AccumulatingTableReader extends PTransform<PBegin, PCollectionTuple> {
+  abstract ImmutableMap<SourceTableReference, PTransform<PBegin, PCollection<SourceRow>>>
+      tableTransforms();
+
+  abstract TupleTag<SourceRow> sourceRowTag();
+
+  abstract TupleTag<SourceTableReference> sourceTableReferenceTag();
+
+  @Override
+  public PCollectionTuple expand(PBegin input) {
+    ImmutableMap<SourceTableReference, PCollection<SourceRow>> tablePCollections =
+        this.tableTransforms().entrySet().stream()
+            .collect(
+                ImmutableMap.toImmutableMap(
+                    Entry::getKey,
+                    e -> input.apply("SourceRowReader." + e.getKey().getName(), e.getValue())));
+    PCollection<SourceRow> sourceRowPCollection =
+        PCollectionList.of(tablePCollections.values())
+            /* Flatten SourceRows read from all the tables into a single PCollection */
+            .apply("SourceRowCollector", Flatten.<SourceRow>pCollections());
+
+    PCollection<SourceTableReference> tableCompletions =
+        PCollectionList.of(
+                tablePCollections.entrySet().stream()
+                    .map(
+                        e ->
+                            /*
+                             * Reduce the Collection of SourceRows to their counts in a parallel stage.
+                             * This ensures that this parallel stage waits till the table is completed,
+                             * allowing the reader to signal completion of reading a table to the pipeline controller.
+                             * As needed, more complex reduce operation can be performed providing additional information to the pipelineController.
+                             * TODO(vardhanvthigle): Add a note on cost vs performance of the Count transform after benchmarking.
+                             */
+                            e.getValue()
+                                .apply("TableRowCounter." + e.getKey().getName(), Count.globally())
```

Comment:
We are creating a named stage per table - does this mean for a 1000 table database, 1000 stages will be created?

I wonder what the graph would look in the Dataflow console - will it be readable? If this will create 2000+ node graph (`TableRowCounter` and `TableCompletion`) it might not render on the console?

Reply:
We will need a stage per table, naming it just makes it a bit easier to debug.

Reply:
Will Add a note on stage per table in the main sections of design (currently there's a very passing reference in alternatives considered)